module/state_synchronization/execution_data_service.go

```diff
@@ -0,0 +1,280 @@
+package state_synchronization
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ipfs/go-cid"
+
+	"github.com/onflow/flow-go/model/encoding"
+	"github.com/onflow/flow-go/network"
+)
+
+const (
+	defaultMaxBlobSize      = 1 << 20 // 1MiB
+	defaultMaxBlobTreeDepth = 2       // prevents malicious CID from causing download of unbounded amounts of data
+	defaultBlobBatchSize    = 16
+)
+
+var ErrBlobTreeDepthExceeded = errors.New("blob tree depth exceeded")
+
+// ExecutionDataService handles adding and getting execution data from a blobservice
+type ExecutionDataService struct {
+	serializer  *serializer
+	blobService network.BlobService
+	maxBlobSize int
+}
+
+func NewExecutionDataService(
+	codec encoding.Codec,
+	compressor network.Compressor,
+	blobService network.BlobService,
+) *ExecutionDataService {
+	return &ExecutionDataService{&serializer{codec, compressor}, blobService, defaultMaxBlobSize}
+}
+
+func (s *ExecutionDataService) receiveBatch(ctx context.Context, br *BlobReceiver) ([]network.Blob, error) {
+	var blobs []network.Blob
+	var err error
+
+	for i := 0; i < defaultBlobBatchSize; i++ {
+		var blob network.Blob
+		blob, err = br.Receive()
+
+		if err != nil {
+			break
+		}
+
+		blobs = append(blobs, blob)
+	}
+
+	return blobs, err
+}
+
+func (s *ExecutionDataService) receiveBlobs(parent context.Context, br *BlobReceiver) ([]cid.Cid, error) {
+	defer br.Close()
+
+	ctx, cancel := context.WithCancel(parent)
+	defer cancel()
+
+	var cids []cid.Cid
+
+	for {
+		batch, recvErr := s.receiveBatch(ctx, br)
+
+		for _, blob := range batch {
+			cids = append(cids, blob.Cid())
+		}
+
+		if err := s.blobService.AddBlobs(ctx, batch); err != nil {
+			return nil, err
+		}
+
+		if recvErr != nil {
+			if recvErr != ErrClosedBlobChannel {
+				// this is an unexpected error, and should never occur
+				return nil, recvErr
+			}
+
+			break
+		}
+	}
+
+	return cids, nil
+}
+
+func (s *ExecutionDataService) addBlobs(ctx context.Context, v interface{}) ([]cid.Cid, error) {
+	bcw, br := IncomingBlobChannel(s.maxBlobSize)
+
+	done := make(chan struct{})
+	var serializeErr error
+
+	go func() {
+		defer close(done)
+		defer bcw.Close()
+
+		if serializeErr = s.serializer.Serialize(bcw, v); serializeErr != nil {
+			return
+		}
+
+		serializeErr = bcw.Flush()
+	}()
+
+	cids, recvErr := s.receiveBlobs(ctx, br)
+
+	<-done
+
+	if recvErr != nil {
+		return nil, recvErr
+	}
+
+	return cids, serializeErr
+}
+
+// Add constructs a blob tree for the given ExecutionData and adds it to the blobservice, and then returns the root CID.
+func (s *ExecutionDataService) Add(ctx context.Context, sd *ExecutionData) (cid.Cid, error) {
+	cids, err := s.addBlobs(ctx, sd)
+
+	if err != nil {
+		return cid.Undef, fmt.Errorf("failed to add execution data blobs: %w", err)
+	}
+
+	for {
+		if len(cids) == 1 {
```

Comment:
how to prevent from endless loop? would `len(cids)` never equal to `1`?

Reply:
`len(cids)` is guaranteed to eventually be equal to 1. It will always get smaller and smaller.